storage: separate command queue for system keys

tbg opened the pull request:
In preparation for #10084 (which will properly account for access to internal
keys in the context of proposer-evaluated KV), split the command queue into a
portion which holds "regular" (=global) reads and writes (such as a key written
to during a SQL insert) and the remainder (=local, though that terminology is
overripe for an overhaul) such as transactional writes to a RangeDescriptor.

The command queue optimizes for certain access patterns by inserting first a
coagulated root span. This optimization would perform poorly when performed
on batches mixing both types of Spans, and it makes no sense for the workload
expected on the "local" portion (where there is high contention with read only
requests and the set of keys touched is small and discrete) and was thus turned
off.

The impact of this is currently negligible, but once the interdependence
between commands is properly accounted for, the "local" command queue will
see a few read-only spans for the majority of spans and its performance will
be of interest, making it a likely target for follow-up optimizations (or
outright specialization) as discussed in #10084.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10173)

<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/keys/keys.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUr-0D7VvdxtD6p9DyM:-KUr-0D7VvdxtD6p9DyN:b-lccxfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/keys/keys.go#L348)):*
> ```Go
> // TODO(tschottdorf): we need a better name for these keys as only some of
> // them are local and it's been identified as an area that is not understood
> // by many of the team's developers. An obvious suggestion is "system" (as
> ```

:+1: "metadata keys" might also provide a similar level of insight as "system keys", although the term might get confused with the meta ranges. You also used "internal" below, which is appropriate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-:-KUr1MkEhqJz8R4Xj-Yd:bg7igkv)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! I'll let this open until @petermattis or @bdarnell have taken a look.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/keys/keys.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUr-0D7VvdxtD6p9DyM:-KUr92enwQU0zznIccMY:b-jjhbx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/keys/keys.go#L348)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> :+1: "metadata keys" might also provide a similar level of insight as "system keys", although the term might get confused with the meta ranges. You also used "internal" below, which is appropriate.
</details>
Yeah, `metadata` is out, but `internal` looks good. I'll see if I can interest someone in a follow-up sweep (if you want to volunteer, be my guest - the naming changes are pretty invasive, so I think it'll take a bit of time to straighten it all out).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-:-KUr8q70wGjV9N8wcxKz:b-k416cb)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PTAL @nvanbenschoten, updated with disabling the covering optimization (I'm cheating a little bit, but avoiding the extra slice item is not trivial).

tamird commented:
Reviewed 3 of 4 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/command_queue.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrRKgn4B7CfDRuBjqb:-KUrRKgn4B7CfDRuBjqc:b-abqz7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1d017d60848492e81f439531c432a051154bb34/pkg/storage/command_queue.go#L104)):*
> ```Go
> 
> // NewCommandQueue returns a new command queue.
> func NewCommandQueue(coveringOptimization bool) *CommandQueue {
> ```

this is not actually being set. do your new tests actually pass?

---

*[pkg/storage/replica.go, line 1156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrP-WxJtaN4c-t_nn0:-KUrP-WygIDrgfsJqKfM:b-d9nwyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/storage/replica.go#L1156)):*
> ```Go
> 	if ba.ReadConsistency != roachpb.INCONSISTENT {
> 		spansGlobal := make([]roachpb.Span, 0, len(ba.Requests))
> 		var spansLocal []roachpb.Span
> ```

a comment to explain the reason for different treatments wouldn't hurt.

---

*[pkg/storage/replica.go, line 1167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrP9uI8wlGKYDGNO_e:-KUrP9uI8wlGKYDGNO_f:b-nss8zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/storage/replica.go#L1167)):*
> ```Go
> 		}
> 
> 		// TODO(tschottdorf): need to optimize read-only status, especially for
> ```

what does it mean to optimize "read-only status"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/command_queue.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrRKgn4B7CfDRuBjqb:-KUrU3FF0vzvURbFLFw_:bzeja1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1d017d60848492e81f439531c432a051154bb34/pkg/storage/command_queue.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is not actually being set. do your new tests actually pass?
</details>
Good catch. My test passes (not setting does the right thing here) but it's bad that the other tests all passed. Added a check that fails them and passing the flag down now.

---

*[pkg/storage/replica.go, line 1156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrP-WxJtaN4c-t_nn0:-KUrUS67ddjzZX408McB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/storage/replica.go#L1156)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> a comment to explain the reason for different treatments wouldn't hurt.
</details>
Done.

---

*[pkg/storage/replica.go, line 1167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrP9uI8wlGKYDGNO_e:-KUrUjR1K47Vk3yFC9se:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/storage/replica.go#L1167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does it mean to optimize "read-only status"?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue_test.go, line 265 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWazOh-DxGTRUH1or:-KUrWazOh-DxGTRUH1os:b-fj0b07) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue_test.go#L265)):*
> ```Go
> 		wk := cq.add(false, a, b)
> 		if act := cq.tree.Len(); act != 1 {
> 			t.Fatal("expected one covering span, but got two spans")
> ```

smells like future rot. you have `act`, why not use it?

---

*[pkg/storage/command_queue_test.go, line 294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWjUYqN32uCpiniJt:-KUrWjUYqN32uCpiniJu:bh1xal) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue_test.go#L294)):*
> ```Go
> 		}
> 		if len(cmd.children) != 2 {
> 			t.Errorf("expected one child in command %+v", cmd)
> ```

this message is not consistent with the check. haven't we learned not to hardcode when we can avoid it?

---

*[pkg/storage/replica.go, line 1184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWyn7CxFWdY0Kpako:-KUrWyn7CxFWdY0Kpakp:b32jmp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/replica.go#L1184)):*
> ```Go
> 
> 		// TODO(tschottdorf): need to make this less global when the local
> 		// command queue is used more heavily as the rare case which has a
> ```

"rare case (split)" instead of having "(splits)" at the end which implies the reads are splits, or something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 4 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 104 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrbhLd9DmXmSDd-S1A:-KUrbhLeUqKjqSwMg80Q:b43lhxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue.go#L104)):*
> ```Go
> 
> // NewCommandQueue returns a new command queue.
> func NewCommandQueue(coveringOptimization bool) *CommandQueue {
> ```

Could you add a comment on why we'd want to turn this optimization on or off, depending on expected workload

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/command_queue.go, line 104 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrbhLd9DmXmSDd-S1A:-KUreysR0R1_zm25dppH:bw9j34h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue.go#L104)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Could you add a comment on why we'd want to turn this optimization on or off, depending on expected workload
</details>
Done, along with an explanation of what the optimization is.

---

*[pkg/storage/command_queue_test.go, line 265 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWazOh-DxGTRUH1or:-KUrc4Y6_MR35Wspk9DQ:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue_test.go#L265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> smells like future rot. you have `act`, why not use it?
</details>
Changed.

---

*[pkg/storage/command_queue_test.go, line 294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWjUYqN32uCpiniJt:-KUrcYm5VttB5bozj5cc:b-yok9b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/command_queue_test.go#L294)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this message is not consistent with the check. haven't we learned not to hardcode when we can avoid it?
</details>
Not my best day. Btw, I'm sure it's not intentional, but that message fails my condescension lint. 

---

*[pkg/storage/replica.go, line 1184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWyn7CxFWdY0Kpako:-KUrdgaXpSgBiNKsqr6i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/replica.go#L1184)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "rare case (split)" instead of having "(splits)" at the end which implies the reads are splits, or something.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 112 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUru0lhqtDxfm1gKj5c:-KUru0lhqtDxfm1gKj5d:bkhp4r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L112)):*
> ```Go
> // the tree.
> // As such, the optimization makes sense for workloads in which commands
> // typically contain many spans, but are spatially disjoint.
> ```

Might want to mention that the covering span optimization is used for "global" keys and disabled for "local" keys.

---

*[pkg/storage/command_queue.go, line 430 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUruVu-_bKefrRAKl0w:-KUruVu-_bKefrRAKl0x:bho5dc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L430)):*
> ```Go
> 	}
> 
> 	if l1, l2 := keys.IsLocal(minKey), keys.IsLocal(maxKey); l1 != l2 {
> ```

Doesn't really matter here, but you're not using `l1` and `l2` below so this could be `keys.IsLocal(minKey) != keys.IsLocal(maxKey)`. 

---

*[pkg/storage/command_queue.go, line 435 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUruc46sigDkoxk_kCh:-KUruc46sigDkoxk_kCi:b636855) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L435)):*
> ```Go
> 			"mixed range-global and range-local keys: %s and %s",
> 			minKey, maxKey,
> 		)
> ```

Not at all a fan of placing the closing paren on a different line. I realize this matches how we format multi-line declarations, but I don't think the same rationale applies here. Feel free to ignore.

---

*[pkg/storage/replica.go, line 1184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWyn7CxFWdY0Kpako:-KUrw2jYWMeWeTvegpOz:b-69rtod) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/replica.go#L1184)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
Isn't #10084 also going to be adding read-only local spans on every request?

---

*[pkg/storage/replica.go, line 258 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUruw5wIAJHF9e-aJEU:-KUruw5wIAJHF9e-aJEV:b-8ouaz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/replica.go#L258)):*
> ```Go
> 		// Enforces at most one command is running per key(s). The global
> 		// component tracks user writes (i.e. all keys for which keys.Addr is
> 		// the identity), the local component the rest (i.e. RangeDescriptor)
> ```

`s/i.e. RangeDescriptor/e.g. RangeDescriptor/g`. Might also mention transaction records.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-:-KUrwJiUrWFH_kTvRA64:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 112 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUru0lhqtDxfm1gKj5c:-KUrxHlcI6GPQBe3NMQS:bk8mu9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L112)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Might want to mention that the covering span optimization is used for "global" keys and disabled for "local" keys.
</details>
+1, though I'd put that comment at the caller to reduce likelihood of rot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: separate command queue for system keys

In preparation for #10084 (which will properly account for access to internal
keys in the context of proposer-evaluated KV), split the command queue into a
portion which holds "regular" (=global) reads and writes (such as a key written
to during a SQL insert) and the remainder (=local, though that terminology is
overripe for an overhaul) such as transactional writes to a RangeDescriptor.

The command queue optimizes for certain access patterns by inserting first a
coagulated root span. This optimization would perform poorly when performed
on batches mixing both types of Spans, and it makes no sense for the workload
expected on the "local" portion (where there is high contention with read only
requests and the set of keys touched is small and discrete) and was thus turned
off.

The impact of this is currently negligible, but once the interdependence
between commands is properly accounted for, the "local" command queue will
see a few read-only spans for the majority of spans and its performance will
be of interest, making it a likely target for follow-up optimizations (or
outright specialization) as discussed in #10084.

```
name                                old time/op    new time/op    delta
Select1_Cockroach-32                  83.6µs ± 3%    83.7µs ± 1%    ~     (p=0.690 n=5+5)
Insert1_Cockroach-32                   503µs ± 1%     506µs ± 1%    ~     (p=0.548 n=5+5)
Insert10_Cockroach-32                  657µs ± 1%     660µs ± 2%    ~     (p=0.310 n=5+5)
Insert100_Cockroach-32                1.65ms ± 1%    1.67ms ± 2%    ~     (p=0.056 n=5+5)
Insert1000_Cockroach-32               10.7ms ± 1%    10.7ms ± 1%    ~     (p=0.421 n=5+5)
InterleavedSelect1000_Cockroach-32    55.0ms ± 1%    55.4ms ± 2%    ~     (p=0.548 n=5+5)

name                                old alloc/op   new alloc/op   delta
Select1_Cockroach-32                  6.33kB ± 0%    6.33kB ± 0%    ~     (p=0.905 n=5+5)
Insert1_Cockroach-32                  34.5kB ± 0%    34.5kB ± 0%  +0.10%  (p=0.032 n=5+5)
Insert10_Cockroach-32                 82.2kB ± 0%    82.2kB ± 0%    ~     (p=0.310 n=5+5)
Insert100_Cockroach-32                 527kB ± 0%     527kB ± 0%    ~     (p=0.548 n=5+5)
Insert1000_Cockroach-32               4.55MB ± 0%    4.55MB ± 0%    ~     (p=0.841 n=5+5)
InterleavedSelect1000_Cockroach-32     797kB ± 0%     797kB ± 0%    ~     (p=0.841 n=5+5)

name                                old allocs/op  new allocs/op  delta
Select1_Cockroach-32                     105 ± 0%       105 ± 0%    ~     (all equal)
Insert1_Cockroach-32                     369 ± 0%       371 ± 0%  +0.54%  (p=0.008 n=5+5)
Insert10_Cockroach-32                    586 ± 0%       588 ± 0%  +0.34%  (p=0.016 n=5+4)
Insert100_Cockroach-32                 2.51k ± 0%     2.52k ± 0%  +0.09%  (p=0.016 n=5+4)
Insert1000_Cockroach-32                21.6k ± 0%     21.6k ± 0%    ~     (p=0.333 n=5+5)
InterleavedSelect1000_Cockroach-32     3.09k ± 0%     3.09k ± 0%  +0.06%  (p=0.048 n=5+4)
```

tschottdorf commented:
Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 112 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUru0lhqtDxfm1gKj5c:-KUryLWNH8Fq1IzIVLJX:b-bj654w) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> +1, though I'd put that comment at the caller to reduce likelihood of rot.
</details>
(Already) done.

---

*[pkg/storage/command_queue.go, line 430 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUruVu-_bKefrRAKl0w:-KUryVLL48h4VqchU1rq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/command_queue.go#L430)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Doesn't really matter here, but you're not using `l1` and `l2` below so this could be `keys.IsLocal(minKey) != keys.IsLocal(maxKey)`. 
</details>
Done.

---

*[pkg/storage/replica.go, line 1184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUrWyn7CxFWdY0Kpako:-KUrysC3JuFyMg9eXFnE:bwxsj0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08f75eca59a205ddf1e3c363028eff98766e76e/pkg/storage/replica.go#L1184)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Isn't #10084 also going to be adding read-only local spans on every request?
</details>
Yes, this is exactly about that PR. Made that clearer by adding an explicit reference.

---

*[pkg/storage/replica.go, line 258 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUruw5wIAJHF9e-aJEU:-KUryozKNufq2f7kylvJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a594a4e341e48eaf3dfa6f614f535c8a86972bd6/pkg/storage/replica.go#L258)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/i.e. RangeDescriptor/e.g. RangeDescriptor/g`. Might also mention transaction records.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 4 files at r1, 1 of 5 files at r4, 1 of 3 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/keys/keys.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173#-KUr-0D7VvdxtD6p9DyM:-KUy8NDnvBJO1hw3FBTt:blqgeyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a661f30f10af7f2d73e898a1eb11cc0f9afab013/pkg/keys/keys.go#L348)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, `metadata` is out, but `internal` looks good. I'll see if I can interest someone in a follow-up sweep (if you want to volunteer, be my guest - the naming changes are pretty invasive, so I think it'll take a bit of time to straighten it all out).
</details>
"Internal" sounds more like range-id-local keys to me. Maybe borrow unicode terminology and call them "planes": "basic plane" and "meta plane"? To me the defining aspect of these keys is the way that they parallel the regular keys. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10173)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Filed https://github.com/cockroachdb/cockroach/issues/10226 for the range-local rename.
